File: discord-logger/src/utils/discordLogger.ts
/**
 * Discord Application Logger
 *
 * Production-grade logger that sends application logs to a Discord channel
 * via webhook. Designed to NEVER crash your application, even if Discord
 * is completely unreachable.
 *
 * Features:
 *  - Batching (sends every 5 seconds to avoid Discord rate limits)
 *  - In-memory queue with overflow protection (configurable max, default 1000)
 *  - Exponential backoff retry logic (3 attempts: 1s, 2s, 4s)
 *  - 5-second timeout per request
 *  - Sensitive data sanitisation (passwords, tokens, API keys, etc.)
 *  - Graceful degradation: console fallback if Discord is unavailable
 *  - Global unhandled exception / promise rejection handlers
 */

'use strict';

// ============================================================
// TYPES & INTERFACES
// ============================================================

export type LogLevel = 'ERROR' | 'WARN' | 'INFO';

export interface LoggerConfig {
  /** Discord webhook URL. Falls back to DISCORD_WEBHOOK_URL env var. */
  webhookUrl?: string;
  /** Identifies which service sent the log (e.g. "RIFF"). */
  serviceName?: string;
  /** e.g. "production" | "development". Falls back to NODE_ENV. */
  environment?: string;
  /** How often (ms) to flush the queue to Discord. Default: 5000. */
  batchInterval?: number;
  /** Max items held in queue before oldest are dropped. Default: 1000. */
  queueLimit?: number;
  /** Set false (or DISCORD_LOGGING_ENABLED=false) to disable Discord sending. */
  enabled?: boolean;
  /** Granular per-level control. */
  logLevels?: {
    error?: boolean;
    warn?: boolean;
    info?: boolean;
  };
}

export interface LogEntry {
  level: LogLevel;
  title: string;
  error?: unknown;
  metadata?: Record<string, unknown>;
  timestamp: string;
}

interface DiscordEmbed {
  title: string;
  description: string;
  color: number;
  timestamp: string;
  footer: { text: string };
  fields?: Array<{ name: string; value: string; inline?: boolean }>;
}

interface DiscordWebhookPayload {
  username: string;
  embeds: DiscordEmbed[];
}

// ============================================================
// CONSTANTS
// ============================================================

/** Discord embed border colours per log level. */
const LOG_COLORS: Record<LogLevel, number> = {
  ERROR: 0xff0000, // Red
  WARN:  0xffa500, // Orange / Yellow
  INFO:  0x0099ff, // Blue
};

/** Visual prefix for each level. */
const LOG_EMOJIS: Record<LogLevel, string> = {
  ERROR: '🔴',
  WARN:  '🟡',
  INFO:  '🔵',
};

const DEFAULT_CONFIG: Required<LoggerConfig> = {
  webhookUrl:    '',
  serviceName:   'RIFF',
  environment:   process.env.NODE_ENV || 'development',
  batchInterval: 5000,
  queueLimit:    1000,
  enabled:       true,
  logLevels:     { error: true, warn: true, info: true },
};

/**
 * Regex patterns that identify sensitive field keys.
 * Matching values are replaced with '[REDACTED]' before being sent to Discord.
 */
const SENSITIVE_PATTERNS = [
  /password/i,
  /secret/i,
  /token/i,
  /api[_-]?key/i,
  /auth/i,
  /credential/i,
  /private/i,
  /webhook/i,
];

// Discord API limits
const DISCORD_MAX_EMBED_DESCRIPTION = 4096;
const DISCORD_MAX_FIELD_VALUE       = 1024;
const DISCORD_MAX_EMBEDS_PER_BATCH  = 10;  // Discord allows max 10 embeds per message
const DISCORD_REQUEST_TIMEOUT_MS    = 5000;
const MAX_RETRY_ATTEMPTS            = 3;

// ============================================================
// UTILITY FUNCTIONS (private, not exported)
// ============================================================

/**
 * Recursively redact values whose key matches a sensitive pattern.
 * Never throws — returns { '[sanitise_error]': '...' } on failure.
 */
function sanitiseData(data: Record<string, unknown>): Record<string, unknown> {
  try {
    const result: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(data)) {
      if (SENSITIVE_PATTERNS.some(p => p.test(key))) {
        result[key] = '[REDACTED]';
      } else if (value !== null && typeof value === 'object' && !Array.isArray(value)) {
        result[key] = sanitiseData(value as Record<string, unknown>);
      } else {
        result[key] = value;
      }
    }
    return result;
  } catch {
    return { '[sanitise_error]': 'Failed to sanitise data' };
  }
}

/** Truncate a string to maxLength, appending a notice when clipped. */
function truncate(str: string, maxLength: number): string {
  if (str.length <= maxLength) return str;
  return str.substring(0, maxLength - 30) + '\n... [truncated]';
}

/** Extract a safe message/stack/code from any thrown value. */
function extractErrorInfo(error: unknown): {
  message: string;
  stack?: string;
  code?: string;
} {
  try {
    if (error instanceof Error) {
      return {
        message: error.message || 'Unknown error',
        stack:   error.stack,
        code:    (error as NodeJS.ErrnoException).code,
      };
    }
    if (typeof error === 'string') return { message: error };
    if (error !== null && typeof error === 'object') {
      const obj = error as Record<string, unknown>;
      return {
        message: String(obj.message ?? obj.error ?? JSON.stringify(error)),
        stack:   typeof obj.stack  === 'string' ? obj.stack  : undefined,
        code:    typeof obj.code   === 'string' ? obj.code   : undefined,
      };
    }
    return { message: String(error) };
  } catch {
    return { message: 'Failed to extract error information' };
  }
}

/** Collect lightweight Node.js process metadata for context. */
function getProcessMetadata(): Record<string, string> {
  try {
    return {
      nodeVersion:   process.version,
      platform:      process.platform,
      arch:          process.arch,
      pid:           String(process.pid),
      uptime:        `${Math.round(process.uptime())}s`,
      memoryUsageMB: `${Math.round(process.memoryUsage().heapUsed / 1024 / 1024)}MB`,
    };
  } catch {
    return {};
  }
}

/** Simple promise-based sleep. */
function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

// ============================================================
// DISCORD LOGGER CLASS
// ============================================================

class DiscordLogger {
  private config: Required<LoggerConfig>;
  private queue: LogEntry[]                           = [];
  private batchTimer: ReturnType<typeof setInterval> | null = null;
  private isFlushing                                  = false;

  constructor(config: LoggerConfig = {}) {
    // Merge defaults, then override with environment variables
    this.config = {
      ...DEFAULT_CONFIG,
      ...config,
      logLevels: { ...DEFAULT_CONFIG.logLevels, ...config.logLevels },
    };

    // Environment variable overrides (env always wins over constructor default,
    // but an explicit constructor value wins over env — mirrored below).
    if (!config.webhookUrl) {
      this.config.webhookUrl = process.env.DISCORD_WEBHOOK_URL || '';
    }
    if (!config.serviceName) {
      this.config.serviceName = process.env.DISCORD_SERVICE_NAME || DEFAULT_CONFIG.serviceName;
    }
    if (config.enabled === undefined) {
      this.config.enabled = process.env.DISCORD_LOGGING_ENABLED !== 'false';
    }
    if (config.batchInterval === undefined) {
      const v = parseInt(process.env.DISCORD_BATCH_INTERVAL || '');
      if (!isNaN(v)) this.config.batchInterval = v;
    }
    if (config.queueLimit === undefined) {
      const v = parseInt(process.env.DISCORD_QUEUE_LIMIT || '');
      if (!isNaN(v)) this.config.queueLimit = v;
    }
  }

  // ============================================================
  // PUBLIC API
  // ============================================================

  /**
   * Re-configure the logger after construction.
   * Safe to call multiple times (e.g. after loading env vars from a config file).
   */
  initialize(config: LoggerConfig): void {
    try {
      this.config = {
        ...this.config,
        ...config,
        logLevels: { ...this.config.logLevels, ...config.logLevels },
      };
      // Always respect the kill-switch env var
      if (process.env.DISCORD_LOGGING_ENABLED === 'false') {
        this.config.enabled = false;
      }
    } catch (err) {
      console.error('[Discord Logger] Failed to initialize:', err);
    }
  }

  /**
   * Start the background batch processor.
   * Call once at application startup — typically right after initialize().
   *
   * @example
   *   discordLogger.startBatchProcessor();
   */
  startBatchProcessor(): void {
    try {
      if (this.batchTimer) this.stopBatchProcessor();

      this.batchTimer = setInterval(() => {
        void this.flushQueue();
      }, this.config.batchInterval);

      // unref() prevents this timer from keeping the process alive on exit
      if (typeof this.batchTimer.unref === 'function') {
        this.batchTimer.unref();
      }

      console.log(`[Discord Logger] Batch processor started (every ${this.config.batchInterval}ms)`);
    } catch (err) {
      console.error('[Discord Logger] Failed to start batch processor:', err);
    }
  }

  /**
   * Stop the batch processor and attempt a final flush of remaining logs.
   * Call during graceful shutdown (SIGTERM, SIGINT).
   */
  stopBatchProcessor(): void {
    try {
      if (this.batchTimer) {
        clearInterval(this.batchTimer);
        this.batchTimer = null;
      }
      if (this.queue.length > 0) {
        console.warn(`[Discord Logger] Stopping with ${this.queue.length} logs in queue — flushing now`);
        void this.flushQueue();
      }
    } catch (err) {
      console.error('[Discord Logger] Failed to stop batch processor:', err);
    }
  }

  /**
   * Register Node.js global handlers for uncaught exceptions and unhandled
   * promise rejections. Call once at the very top of your entry point.
   */
  registerGlobalHandlers(): void {
    try {
      process.on('uncaughtException', (error: Error) => {
        console.error('[Discord Logger] Uncaught Exception:', error);
        this.addToQueue('ERROR', 'Uncaught Exception', {
          error,
          fatal:       true,
          processInfo: getProcessMetadata(),
        });
        void this.flushQueue(); // best-effort immediate send
      });

      process.on('unhandledRejection', (reason: unknown) => {
        console.error('[Discord Logger] Unhandled Promise Rejection:', reason);
        this.addToQueue('ERROR', 'Unhandled Promise Rejection', {
          error:       reason,
          processInfo: getProcessMetadata(),
        });
        void this.flushQueue();
      });

      console.log('[Discord Logger] Global error handlers registered');
    } catch (err) {
      console.error('[Discord Logger] Failed to register global handlers:', err);
    }
  }

  // ----------------------------------------------------------
  // Log level methods
  // ----------------------------------------------------------

  /**
   * Log a critical error (red embed).
   *
   * @param title   Short description of what failed.
   * @param context Optional object. Include `error` key for Error/exception.
   *
   * @example
   *   discordLogger.error('WooCommerce Sync Failed', { error, orderId: 123 });
   */
  error(title: string, context?: { error?: unknown; [key: string]: unknown }): void {
    if (!this.config.logLevels.error) return;
    this.addToQueue('ERROR', title, context);
  }

  /**
   * Log a warning (yellow/orange embed).
   *
   * @example
   *   discordLogger.warn('Low stock detected', { sku: 'REV-456', qty: 1 });
   */
  warn(title: string, context?: { error?: unknown; [key: string]: unknown }): void {
    if (!this.config.logLevels.warn) return;
    this.addToQueue('WARN', title, context);
  }

  /**
   * Log informational data (blue embed).
   *
   * @example
   *   discordLogger.info('Deployment complete', { version: '2.1.0' });
   */
  info(title: string, context?: { error?: unknown; [key: string]: unknown }): void {
    if (!this.config.logLevels.info) return;
    this.addToQueue('INFO', title, context);
  }

  // ============================================================
  // PRIVATE METHODS
  // ============================================================

  /**
   * Add an entry to the in-memory queue.
   * Drops the oldest entry (with a warning) when the queue is full.
   * Always writes to console as a fallback — even when Discord is disabled.
   */
  private addToQueue(
    level: LogLevel,
    title: string,
    context?: { error?: unknown; [key: string]: unknown },
  ): void {
    try {
      // Console fallback regardless of Discord status
      const logFn = level === 'ERROR'
        ? console.error
        : level === 'WARN'
        ? console.warn
        : console.info;
      logFn(`[${level}] ${title}`, context ?? '');

      if (!this.config.enabled) return;
      if (!this.config.webhookUrl) {
        console.warn(`[Discord Logger] No webhook URL configured — skipping Discord send for: [${level}] ${title}`);
        return;
      }

      // Separate the error from general metadata
      const { error, ...metadata } = context ?? {};

      // Drop oldest when at capacity
      if (this.queue.length >= this.config.queueLimit) {
        const dropped = this.queue.shift();
        console.warn(
          `[Discord Logger] Queue full (${this.config.queueLimit}). Dropped: [${dropped?.level}] ${dropped?.title}`
        );
      }

      this.queue.push({
        level,
        title,
        error,
        metadata: Object.keys(metadata).length > 0
          ? (metadata as Record<string, unknown>)
          : undefined,
        timestamp: new Date().toISOString(),
      });
    } catch (err) {
      // Absolute last resort — never propagate
      console.error('[Discord Logger] addToQueue failed:', err, `| Original: [${level}] ${title}`);
    }
  }

  /**
   * Drain the queue and send all pending entries to Discord.
   * Called automatically by the batch timer, and manually on shutdown/errors.
   */
  private async flushQueue(): Promise<void> {
    if (this.isFlushing || this.queue.length === 0) return;
    this.isFlushing = true;

    try {
      // Drain the entire queue atomically (new entries added while we're
      // sending will be picked up on the next flush cycle).
      const batch = this.queue.splice(0, this.queue.length);

      // Build Discord embeds — skip any that fail to build
      const embeds: DiscordEmbed[] = [];
      for (const entry of batch) {
        try {
          embeds.push(this.buildEmbed(entry));
        } catch (err) {
          console.error('[Discord Logger] Failed to build embed for entry:', entry.title, err);
        }
      }

      // Discord allows max 10 embeds per webhook call — chunk if needed
      for (let i = 0; i < embeds.length; i += DISCORD_MAX_EMBEDS_PER_BATCH) {
        const chunk = embeds.slice(i, i + DISCORD_MAX_EMBEDS_PER_BATCH);
        await this.sendToDiscord({
          username: `${this.config.serviceName} Logger`,
          embeds:   chunk,
        });
      }
    } catch (err) {
      console.error('[Discord Logger] flushQueue failed:', err);
    } finally {
      this.isFlushing = false;
    }
  }

  /**
   * Convert a LogEntry into a Discord embed object.
   */
  private buildEmbed(entry: LogEntry): DiscordEmbed {
    const errorInfo = entry.error ? extractErrorInfo(entry.error) : null;
    const fields: NonNullable<DiscordEmbed['fields']> = [];

    // --- Error details ---
    if (errorInfo) {
      if (errorInfo.code) {
        fields.push({
          name:   'Error Code',
          value:  truncate(errorInfo.code, DISCORD_MAX_FIELD_VALUE),
          inline: true,
        });
      }
      if (errorInfo.stack) {
        fields.push({
          name:   'Stack Trace',
          value:  `\`\`\`\n${truncate(errorInfo.stack, DISCORD_MAX_FIELD_VALUE - 10)}\n\`\`\``,
          inline: false,
        });
      }
    }

    // --- Sanitised metadata fields ---
    if (entry.metadata) {
      const safe = sanitiseData(entry.metadata);

      for (const [key, value] of Object.entries(safe)) {
        // processInfo gets its own dedicated field below
        if (key === 'processInfo') continue;

        const str = typeof value === 'object'
          ? JSON.stringify(value, null, 2)
          : String(value ?? '');

        fields.push({
          name:   key,
          value:  truncate(str, DISCORD_MAX_FIELD_VALUE),
          inline: typeof value !== 'object',
        });
      }

      // Condense processInfo into a single neat field
      if (safe.processInfo && typeof safe.processInfo === 'object') {
        const info = safe.processInfo as Record<string, string>;
        fields.push({
          name:   'Process Info',
          value:  truncate(
            Object.entries(info).map(([k, v]) => `**${k}:** ${v}`).join('\n'),
            DISCORD_MAX_FIELD_VALUE
          ),
          inline: false,
        });
      }
    }

    // --- Embed description ---
    let description = '';
    if (errorInfo?.message) {
      description = `**${truncate(errorInfo.message, 512)}**`;
    }

    return {
      title:       `${LOG_EMOJIS[entry.level]} ${entry.level} — ${truncate(entry.title, 200)}`,
      description: truncate(description, DISCORD_MAX_EMBED_DESCRIPTION),
      color:       LOG_COLORS[entry.level],
      timestamp:   entry.timestamp,
      footer:      { text: `${this.config.serviceName} | ${this.config.environment}` },
      fields:      fields.slice(0, 25), // Discord hard limit: 25 fields per embed
    };
  }

  /**
   * POST a webhook payload to Discord with:
   *  - 5-second request timeout
   *  - Rate-limit (429) awareness
   *  - Exponential backoff: 1s → 2s → 4s
   */
  private async sendToDiscord(payload: DiscordWebhookPayload): Promise<void> {
    if (!this.config.webhookUrl) return;

    for (let attempt = 1; attempt <= MAX_RETRY_ATTEMPTS; attempt++) {
      try {
        const controller = new AbortController();
        const timeoutId  = setTimeout(() => controller.abort(), DISCORD_REQUEST_TIMEOUT_MS);

        const response = await fetch(this.config.webhookUrl, {
          method:  'POST',
          headers: { 'Content-Type': 'application/json' },
          body:    JSON.stringify(payload),
          signal:  controller.signal,
        });

        clearTimeout(timeoutId);

        // Handle rate limiting
        if (response.status === 429) {
          const retryAfterSec = parseInt(response.headers.get('retry-after') ?? '5', 10);
          const delayMs       = retryAfterSec * 1000;
          console.warn(`[Discord Logger] Rate limited — waiting ${delayMs}ms before retry`);
          await sleep(delayMs);
          continue; // retry without decrementing attempt counter would be more correct,
                    // but to keep things simple we count this as an attempt
        }

        if (!response.ok) {
          const body = await response.text().catch(() => '<unreadable>');
          throw new Error(`Discord webhook HTTP ${response.status}: ${body}`);
        }

        console.log(`[Discord Logger] Batch sent: ${payload.embeds.length} embed(s)`);
        return; // success

      } catch (err) {
        if (attempt === MAX_RETRY_ATTEMPTS) {
          console.error(
            `[Discord Logger] Failed after ${MAX_RETRY_ATTEMPTS} attempts. Dropping batch.`,
            err
          );
          return;
        }

        // Exponential backoff: 1 000ms, 2 000ms, 4 000ms
        const backoffMs = Math.pow(2, attempt - 1) * 1000;
        console.warn(
          `[Discord Logger] Send failed (attempt ${attempt}/${MAX_RETRY_ATTEMPTS}). Retrying in ${backoffMs}ms…`,
          err
        );
        await sleep(backoffMs);
      }
    }
  }
}

// ============================================================
// SINGLETON EXPORT
// ============================================================

/**
 * Pre-configured singleton instance.
 *
 * Reads DISCORD_WEBHOOK_URL, DISCORD_SERVICE_NAME, DISCORD_LOGGING_ENABLED,
 * DISCORD_BATCH_INTERVAL, and DISCORD_QUEUE_LIMIT from environment variables
 * automatically.
 *
 * @example
 *   import { discordLogger } from './utils/discordLogger';
 *
 *   discordLogger.initialize({ serviceName: 'RIFF', environment: 'production' });
 *   discordLogger.startBatchProcessor();
 *   discordLogger.error('Sync failed', { error, orderId: 42 });
 */
export const discordLogger = new DiscordLogger();

export { DiscordLogger };
export default discordLogger;

File: inventory_system/app/core/discord_logger.py
"""
app/core/discord_logger.py

A Python logging.Handler that forwards WARNING+ log records to Discord
via a webhook. Designed to be non-blocking and to never crash the app.

Key properties:
  - Subclasses logging.Handler so every logger.error() / logger.warning()
    in the codebase is automatically forwarded — no code changes elsewhere.
  - Batches records and sends them every `batch_interval` seconds via an
    asyncio background task to avoid Discord's rate limits.
  - In-memory queue (deque) with configurable max size. When full the oldest
    entry is dropped silently (Python's deque(maxlen=N) behaviour).
  - Exponential backoff retry: up to 3 attempts (1 s → 2 s → 4 s).
  - 5-second hard timeout on every HTTP request.
  - Sensitive field names (password, token, api_key …) are redacted before
    being sent to Discord.
  - Never raises an exception — all errors are written to stderr/console as
    a last-resort fallback.

Usage (see app/main.py lifespan for the real wiring):

    handler = DiscordLogHandler(webhook_url=..., service_name="RIFF")
    logging.getLogger().addHandler(handler)
    handler.start_batch_processor(asyncio.get_event_loop())
    # on shutdown:
    handler.stop_batch_processor()
    await handler.flush_now()
"""

from __future__ import annotations

import asyncio
import logging
import os
import traceback
from collections import deque
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

import httpx

# ---------------------------------------------------------------------------
# Module-level logger (for internal diagnostics — goes to console only)
# ---------------------------------------------------------------------------
_log = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Discord embed constants
# ---------------------------------------------------------------------------

# Embed border colour per Python log level name
_COLOURS: Dict[str, int] = {
    "CRITICAL": 0x8B0000,  # Dark red
    "ERROR":    0xFF0000,  # Red
    "WARNING":  0xFFA500,  # Orange
    "INFO":     0x0099FF,  # Blue
    "DEBUG":    0x808080,  # Grey
}

_EMOJIS: Dict[str, str] = {
    "CRITICAL": "🔴",
    "ERROR":    "🔴",
    "WARNING":  "🟡",
    "INFO":     "🔵",
    "DEBUG":    "⚪",
}

# Discord API limits
_MAX_EMBED_DESCRIPTION = 4096
_MAX_FIELD_VALUE       = 1024
_MAX_EMBEDS_PER_MSG    = 10   # Discord hard cap
_REQUEST_TIMEOUT       = 5.0  # seconds
_MAX_RETRIES           = 3

# Substrings that mark a metadata key as sensitive (case-insensitive match)
_SENSITIVE_SUBSTRINGS = (
    "password", "secret", "token", "api_key", "apikey",
    "auth", "credential", "private", "webhook",
)

# ---------------------------------------------------------------------------
# Utilities
# ---------------------------------------------------------------------------

def _truncate(text: str, max_len: int) -> str:
    """Return text clipped to max_len chars with a notice if clipped."""
    if len(text) <= max_len:
        return text
    return text[: max_len - 30] + "\n… [truncated]"


def _sanitise(data: Dict[str, Any]) -> Dict[str, Any]:
    """
    Return a copy of `data` with values redacted when the key contains a
    sensitive substring.  Recurses into nested dicts.  Never raises.
    """
    try:
        result: Dict[str, Any] = {}
        for key, value in data.items():
            lower = key.lower()
            if any(s in lower for s in _SENSITIVE_SUBSTRINGS):
                result[key] = "[REDACTED]"
            elif isinstance(value, dict):
                result[key] = _sanitise(value)
            else:
                result[key] = value
        return result
    except Exception:
        return {"[sanitise_error]": "Failed to sanitise data"}


# ---------------------------------------------------------------------------
# Main handler class
# ---------------------------------------------------------------------------

class DiscordLogHandler(logging.Handler):
    """
    A logging.Handler that queues Python log records and sends them to a
    Discord channel via webhook on a configurable batch interval.

    Constructor arguments all fall back to environment variables so you can
    configure it purely through Railway/your .env file:

        DISCORD_WEBHOOK_URL       — webhook URL (required)
        DISCORD_SERVICE_NAME      — label in every embed footer (default: RIFF)
        DISCORD_LOGGING_ENABLED   — set "false" to disable entirely
        DISCORD_BATCH_INTERVAL    — seconds between flushes (default: 5)
        DISCORD_QUEUE_LIMIT       — max queued records (default: 1000)
    """

    def __init__(
        self,
        webhook_url:    Optional[str] = None,
        service_name:   Optional[str] = None,
        environment:    Optional[str] = None,
        batch_interval: Optional[float] = None,
        queue_limit:    Optional[int] = None,
        min_level:      int = logging.WARNING,
    ) -> None:
        super().__init__(level=min_level)

        # Resolve config: constructor argument → environment variable → default
        self.webhook_url = (
            webhook_url
            or os.getenv("DISCORD_WEBHOOK_URL", "")
        )
        self.service_name = (
            service_name
            or os.getenv("DISCORD_SERVICE_NAME", "RIFF")
        )
        self.environment = (
            environment
            or os.getenv("ENVIRONMENT", os.getenv("NODE_ENV", "production"))
        )
        self.batch_interval = float(
            batch_interval
            if batch_interval is not None
            else os.getenv("DISCORD_BATCH_INTERVAL", "5")
        )
        self.queue_limit = int(
            queue_limit
            if queue_limit is not None
            else os.getenv("DISCORD_QUEUE_LIMIT", "1000")
        )
        self.enabled: bool = (
            os.getenv("DISCORD_LOGGING_ENABLED", "true").lower() != "false"
        )

        # deque(maxlen=N) automatically drops the oldest entry when full —
        # no manual overflow logic needed.
        self._queue: deque[Dict[str, Any]] = deque(maxlen=self.queue_limit)

        self._batch_task: Optional[asyncio.Task[None]] = None

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def start_batch_processor(self) -> None:
        """
        Create the asyncio background task that flushes the queue every
        `batch_interval` seconds.

        Call this from inside an async context (e.g. FastAPI lifespan)
        so that asyncio.get_event_loop() / create_task() works correctly.
        """
        try:
            loop = asyncio.get_event_loop()
            self._batch_task = loop.create_task(
                self._batch_loop(),
                name="discord-logger-batch",
            )
            _log.info(
                "[Discord Logger] Batch processor started "
                "(interval=%.0fs, queue_limit=%d, enabled=%s)",
                self.batch_interval, self.queue_limit, self.enabled,
            )
        except Exception as exc:
            # Non-fatal — log to console and continue
            print(f"[Discord Logger] Failed to start batch processor: {exc}")

    def stop_batch_processor(self) -> None:
        """
        Cancel the batch task.  The CancelledError handler inside
        _batch_loop() will attempt one final flush before exiting.

        Call this from the FastAPI lifespan shutdown block.
        """
        if self._batch_task and not self._batch_task.done():
            self._batch_task.cancel()
        _log.info("[Discord Logger] Batch processor stopped.")

    async def flush_now(self) -> None:
        """Force an immediate send of all queued records. Useful on shutdown."""
        await self._send_batch()

    # ------------------------------------------------------------------
    # logging.Handler interface
    # ------------------------------------------------------------------

    def emit(self, record: logging.LogRecord) -> None:
        """
        Called by Python's logging system for every matching log record.
        Queues the record for the next batch flush.
        This method must never raise — errors are handled by handleError().
        """
        try:
            if not self.enabled or not self.webhook_url:
                return

            # Format the full message (includes exception text if present)
            try:
                full_message = self.format(record)
            except Exception:
                full_message = record.getMessage()

            # Extract a short title from the raw message (no exc info)
            title = record.getMessage()

            # Capture traceback separately for the embed Stack Trace field
            exc_text: Optional[str] = None
            if record.exc_info and record.exc_info[0] is not None:
                exc_text = "".join(traceback.format_exception(*record.exc_info))

            entry: Dict[str, Any] = {
                "level":       record.levelname,
                "title":       title[:200],
                "message":     full_message,
                "logger_name": record.name,
                "timestamp":   datetime.fromtimestamp(
                    record.created, tz=timezone.utc
                ).isoformat(),
                "exc_text":    exc_text,
            }

            self._queue.append(entry)

        except Exception:
            # Last resort — logging.Handler.handleError writes to sys.stderr
            self.handleError(record)

    # ------------------------------------------------------------------
    # Private: batch loop
    # ------------------------------------------------------------------

    async def _batch_loop(self) -> None:
        """
        Background asyncio task.  Sleeps for batch_interval then flushes.
        On CancelledError (graceful shutdown) attempts a final flush.
        """
        while True:
            try:
                await asyncio.sleep(self.batch_interval)
                await self._send_batch()
            except asyncio.CancelledError:
                # Shutdown path — best-effort final flush
                try:
                    await self._send_batch()
                except Exception as exc:
                    print(f"[Discord Logger] Final flush failed: {exc}")
                raise  # re-raise so the task actually ends
            except Exception as exc:
                # Never crash the loop — log and continue
                print(f"[Discord Logger] Batch loop error (will retry next cycle): {exc}")

    # ------------------------------------------------------------------
    # Private: build and send
    # ------------------------------------------------------------------

    async def _send_batch(self) -> None:
        """
        Atomically drain the queue, build Discord embeds, and POST them
        in chunks of up to 10 (Discord's per-message limit).
        """
        if not self._queue:
            return

        try:
            # Drain the queue atomically — records added while we're
            # sending are picked up on the next cycle.
            batch: List[Dict[str, Any]] = list(self._queue)
            self._queue.clear()

            embeds: List[Dict[str, Any]] = []
            for entry in batch:
                try:
                    embeds.append(self._build_embed(entry))
                except Exception as exc:
                    print(f"[Discord Logger] Failed to build embed for '{entry.get('title')}': {exc}")

            # Send in chunks of 10
            for i in range(0, len(embeds), _MAX_EMBEDS_PER_MSG):
                chunk = embeds[i : i + _MAX_EMBEDS_PER_MSG]
                await self._post_to_discord({
                    "username": f"{self.service_name} Logger",
                    "embeds":   chunk,
                })

        except Exception as exc:
            print(f"[Discord Logger] _send_batch failed: {exc}")

    def _build_embed(self, entry: Dict[str, Any]) -> Dict[str, Any]:
        """Convert a queued log-entry dict to a Discord embed dict."""
        level   = entry.get("level", "INFO")
        colour  = _COLOURS.get(level, 0x808080)
        emoji   = _EMOJIS.get(level, "⚪")
        title   = entry.get("title", "Log entry")
        message = entry.get("message", title)

        fields: List[Dict[str, Any]] = []

        # Logger name
        logger_name = entry.get("logger_name", "")
        if logger_name:
            fields.append({
                "name":   "Logger",
                "value":  _truncate(logger_name, 256),
                "inline": True,
            })

        # Stack trace
        exc_text = entry.get("exc_text")
        if exc_text:
            fields.append({
                "name":   "Stack Trace",
                "value":  f"```\n{_truncate(exc_text, _MAX_FIELD_VALUE - 10)}\n```",
                "inline": False,
            })

        # Description: full formatted message (may include exc repr line)
        description = _truncate(message, _MAX_EMBED_DESCRIPTION)

        return {
            "title":       f"{emoji} {level} — {_truncate(title, 200)}",
            "description": description,
            "color":       colour,
            "timestamp":   entry.get("timestamp"),
            "footer":      {"text": f"{self.service_name} | {self.environment}"},
            "fields":      fields[:25],  # Discord hard limit: 25 fields
        }

    async def _post_to_discord(self, payload: Dict[str, Any]) -> None:
        """
        POST `payload` to the Discord webhook.
        Retries up to _MAX_RETRIES times with exponential backoff.
        Handles rate-limiting (HTTP 429) using the retry-after header.
        Times out after _REQUEST_TIMEOUT seconds.
        Never raises.
        """
        if not self.webhook_url:
            return

        for attempt in range(1, _MAX_RETRIES + 1):
            try:
                async with httpx.AsyncClient(timeout=_REQUEST_TIMEOUT) as client:
                    response = await client.post(
                        self.webhook_url,
                        json=payload,
                    )

                # Rate limited — wait as Discord instructs
                if response.status_code == 429:
                    retry_after = float(
                        response.headers.get("retry-after", "5")
                    )
                    print(
                        f"[Discord Logger] Rate limited — waiting {retry_after:.1f}s "
                        f"(attempt {attempt}/{_MAX_RETRIES})"
                    )
                    await asyncio.sleep(retry_after)
                    continue

                if response.status_code not in (200, 204):
                    raise httpx.HTTPStatusError(
                        f"Discord returned {response.status_code}: "
                        f"{response.text[:200]}",
                        request=response.request,
                        response=response,
                    )

                # Success
                print(
                    f"[Discord Logger] Sent {len(payload['embeds'])} embed(s) "
                    f"to #{self.service_name}"
                )
                return

            except Exception as exc:
                if attempt == _MAX_RETRIES:
                    print(
                        f"[Discord Logger] Failed after {_MAX_RETRIES} attempts — "
                        f"dropping batch. Last error: {exc}"
                    )
                    return

                # Exponential backoff: 1s, 2s, 4s
                backoff = 2 ** (attempt - 1)
                print(
                    f"[Discord Logger] Attempt {attempt}/{_MAX_RETRIES} failed "
                    f"({exc}). Retrying in {backoff}s…"
                )
                await asyncio.sleep(backoff)
